cpp/src/memory_handle.cpp

```diff
@@ -0,0 +1,81 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <cstdio>
+#include <memory>
+#include <string>
+
+#include <ucp/api/ucp.h>
+
+#include <ucs/memory/memory_type.h>
+#include <ucxx/constructors.h>
+#include <ucxx/log.h>
+#include <ucxx/memory_handle.h>
+#include <ucxx/utils/ucx.h>
+
+namespace ucxx {
+
+MemoryHandle::MemoryHandle(std::shared_ptr<Context> context, const size_t size, void* buffer)
+{
+  setParent(context);
+
+  ucp_mem_map_params_t params;
+  if (buffer == nullptr) {
+    params = {.field_mask  = UCP_MEM_MAP_PARAM_FIELD_FLAGS | UCP_MEM_MAP_PARAM_FIELD_LENGTH,
+              .length      = size,
+              .flags       = UCP_MEM_MAP_NONBLOCK | UCP_MEM_MAP_ALLOCATE,
+              .memory_type = UCS_MEMORY_TYPE_HOST};
+  } else {
+    params = {.field_mask  = UCP_MEM_MAP_PARAM_FIELD_ADDRESS | UCP_MEM_MAP_PARAM_FIELD_LENGTH,
+              .address     = buffer,
+              .length      = size,
+              .memory_type = UCS_MEMORY_TYPE_HOST};
```

Comment:
*question*: Memory handles only work with host memory? Or we're only supporting host memory right now?